devicemodel/samples/nuc/launch_xenomai.sh

```diff
@@ -0,0 +1,219 @@
+		echo dev_bus:$dev_bus
+		echo dev_dev:$dev_dev
+		echo dev_fun:$dev_fun
+		echo dev_drv:$dev_drv
+		echo dev_passthru_opt:$dev_passthru_opt
+
+		# replace its driver with pci-stub
+		echo "$dev_vendor_id $dev_device_id" > /sys/bus/pci/drivers/pci-stub/new_id
+		echo $dev_bdf > $dev_path/driver/unbind
+		echo $dev_bdf > /sys/bus/pci/drivers/pci-stub/bind
+
+		# add to global options
+		dev_opts="$dev_opts $dev_passthru_opt"
+
+		# register for de-passthrough later
+		passthru_devs[$dev_bdf]=$dev_drv
+
+		# increase virtual pci slot number
+		guest_pci_nr=$(expr $guest_pci_nr + 1)
+	done
+}
+
+function de_passthru_pci_devs()
+{
+	for dev_bdf in "${!passthru_devs[@]}"
+	do
+		echo $dev_bdf > /sys/bus/pci/drivers/pci-stub/unbind
+		echo $dev_bdf > /sys/bus/pci/drivers/${passthru_devs[$dev_bdf]}/bind
+	done
+}
+
+# restore pci device driver on exit
+trap de_passthru_pci_devs EXIT
+
+mem_size=1024M
+eth=
+disk=
+img=
+kernel=
+passthrus=()
+while getopts hn:b:f:k:B: opt
+do
+	case "${opt}" in
+		h)
+			usage;
+			;;
+		n)
+			eth=${OPTARG}
+			passthrus+=($eth)
+			;;
+		b)
+			disk="${OPTARG}"
+			passthrus+=($disk)
+			;;
+		f)
+			img="${OPTARG}"
+			;;
+		k)
+			kernel="${OPTARG}"
+			;;
+		B)
+			mem_size="${OPTARG}"
+			;;
+		?)
+			echo "{OPTARG}"
+			;;
+	esac
+done
+
+if [[ ( -z "${kernel}") || (( -z "${disk}") && ( -z "${img}" )) ]]; then
+    usage; exit 1
+fi
+
+# create a unique tap device for each VM
+if [ -z $eth ];then
+	tap=tap_rtvm
+	tap_exist=$(ip a | grep "$tap" | awk '{print $1}')
+	if [ "$tap_exist"x != "x" ]; then
+		echo "tap device existed, reuse $tap"
+	else
+		ip tuntap add dev $tap mode tap
+		ip link set "$tap" master acrn-br0
+		ip link set dev "$tap" down
+		ip link set dev "$tap" up
+	fi
+
+	dev_opts="$dev_opts -s $guest_pci_nr,virtio-net,$tap"
+	guest_pci_nr=$(expr $guest_pci_nr + 1)
+fi
+
+function launch_hard_rt_vm()
+{
+mkdir -p /mnt
+if [ ! -z $img ]; then
+	# mount virtual disk (assuming the last partition is root)
+	vdisk=$(losetup -f -P --show $img)
+	vdisk_root=$(ls -1 ${vdisk}p* | tail -1)
+	mount $vdisk_root /mnt
+
+	# extract the kernel and install kernel modules
+	bzImage=/mnt/$(tar xzvf $kernel -C /mnt --dereference --exclude=vmlinux* | grep vmlinuz)
+	cp $bzImage /tmp/bzImage.xenomai
+
+	# umount disk
+	umount /mnt
+	losetup -d $vdisk
+
+	# find appropriate root partition information
+	disk_part_nr=$(basename $vdisk_root| sed 's/.*p\([0-9]$\)/\1/')
+	vdisk=/dev/vda$disk_part_nr
+
+	dev_opts="$dev_opts -s $guest_pci_nr,virtio-blk,$img"
+	guest_pci_nr=$(expr $guest_pci_nr + 1)
+elif [ ! -z $disk ];then
+	# mount the physical disk
+	mount $disk /mnt
+
+	# extract the kernel and install kernel modules
+	bzImage=/mnt/$(tar xzvf $kernel -C /mnt --dereference --exclude=vmlinux* | grep vmlinuz)
+	cp $bzImage /tmp/bzImage.xenomai
+
+	# umount disk
+	umount /mnt
+
+	# detect partition information
+	disk_part_nr=$(basename $disk| sed 's/.*\([0-9]$\)/\1/')
+	if grep -q "nvme" "$disk"; then
```

Comment:
should we also account for people usign eMMC (i.e. `mmcblk`)?

Reply:
Yes, support for more types of storage (e.g. eMMC) will be great. As this is just a sample script for users to refer to and NVMe and SATA are two major types we support now. Let's keep it simple for now, maybe we could add more later.